src/module/service.cpp

```diff
@@ -0,0 +1,290 @@
+    }
+
+    if (module->services.find(cfg.api) == module->services.end()) {
+        module->lock.unlock();
+        return grpc::Status(grpc::UNKNOWN, "module cannot service api " + cfg.api.to_string());
+    }
+
+    std::shared_ptr<SubtypeService> sub_svc = module->services.at(cfg.api);
+    if (cfg.api.is_component_type() && !(cfg.api == Generic::subtype())) {
+        if (module->services.find(Generic::subtype()) == module->services.end()) {
+            module->lock.unlock();
+            return grpc::Status(grpc::UNKNOWN, "module cannot service the generic api");
+        }
+        std::shared_ptr<SubtypeService> generic_service = module->services.at(Generic::subtype());
+        generic_service->add(cfg.resource_name(), res);
+    }
+    sub_svc->add(cfg.resource_name(), res);
+
+    module->lock.unlock();
+    return grpc::Status();
+};
+
+::grpc::Status ModuleService_::ReconfigureResource(
+    ::grpc::ServerContext* context,
+    const ::viam::module::v1::ReconfigureResourceRequest* request,
+    ::viam::module::v1::ReconfigureResourceResponse* response) {
+    viam::app::v1::ComponentConfig proto = request->config();
+    Component cfg = Component::from_proto(proto);
+    std::shared_ptr<Module> module = this->module;
+
+    Dependencies deps = get_dependencies(this, request->dependencies());
+
+    if (module->services.find(cfg.api) == module->services.end()) {
+        return grpc::Status(grpc::UNKNOWN, "no rpc service for config: " + cfg.api.to_string());
+    }
+    std::shared_ptr<SubtypeService> sub_svc = module->services.at(cfg.api);
+
+    // see if our resource is reconfigurable. if it is, reconfigure
+    std::shared_ptr<ResourceBase> res = sub_svc->resource(cfg.resource_name().name);
+    if (res == nullptr) {
+        return grpc::Status(
+            grpc::UNKNOWN,
+            "unable to reconfigure resource " + cfg.resource_name().name + " as it doesn't exist.");
+    }
+    try {
+        res->reconfigure(deps);
+        return grpc::Status();
+    } catch (std::exception& exc) {
+    }
+
+    // if the type isn't reconfigurable by default, replace it
+    try {
+        res->stop();
+    } catch (std::string err) {
+        BOOST_LOG_TRIVIAL(error) << "unable to stop resource: " << err;
+    }
+
+    if (cfg.api.is_component_type()) {
+        std::shared_ptr<ComponentRegistration> reg = Registry::lookup_component(cfg.name);
+        if (reg != nullptr) {
+            std::shared_ptr<ComponentBase> comp =
+                reg->create_rpc_client(module->name, module->channel);
+            if (!(cfg.api == Generic::subtype())) {
+                if (module->services.find(Generic::subtype()) == module->services.end()) {
+                    return grpc::Status(grpc::UNKNOWN, "no generic service");
+                }
+                std::shared_ptr<SubtypeService> generic_service =
+                    module->services.at(Generic::subtype());
+                generic_service->replace_one(cfg.resource_name(), res);
```

Comment:
probably update this logic as well because we don't need generic service to serve everything now

Reply:
handling in subsequent pr.